### PR TITLE
feat(EC-1957): verify base image release signatures

### PR DESCRIPTION
**Pull request by joejstuart**
## Summary

- Add signature-based base image verification using `ec.sigstore.verify_image()` with `signing_identities` rule data (named entry `rh-release`)
- Deprecation warning when `allowed_registry_prefixes` is configured without `signing_identities`
- Validation: `signing_identities` structure, entry types, sigstore opts (via `sigstore.validate`), and missing `rh-release` key
- `allowed_registry_prefixes` format always validated (type, uniqueness); `minItems` requirement relaxed when a valid signing identity is present
- Comprehensive tests covering signature verification, deprecation warnings, validation, and edge cases

Resolves: EC-1957

## Test plan

- [x] `make ci` passes (1025/1025 tests, 0 lint violations)
- [ ] Review signature verification with a real key against a signed image
- [ ] Verify deprecation warning appears in policy evaluation output

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

File: policy/release/base_image_registries/base_image_registries.rego [hunk 1]
```diff
@@ -170,8 +181,8 @@ _cyclonedx_image_ref(component) := image_ref if {
 	image_ref := sbom.image_ref_from_purl(purl)
 }
 
-# Verify allowed_registry_prefixes is a non-empty list of strings
 _rule_data_errors contains error if {
+	not _release_public_key
 	some e in j.validate_schema(
```
Comment (joejstuart) [inline policy/release/base_image_registries/base_image_registries.rego:185]: Not fixing — `allowed_registry_prefixes` is deprecated and will be removed in a future release. Adding more validation for a field we're phasing out isn't worth the complexity.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 2]
```diff
@@ -188,4 +199,28 @@ _rule_data_errors contains error if {
...
+	count(prefixes) > 0
+	not _release_public_key
+	error := {
+		# regal ignore:line-length
+		"message": "allowed_registry_prefixes is configured without release_public_key. Migrate to signature-based verification by setting release_public_key in rule data.",
+		"severity": "warning",
+	}
+}
+
+_rule_data_errors contains error if {
```
Comment (joejstuart) [inline policy/release/base_image_registries/base_image_registries.rego:213]: Not applicable — `j.validate_schema` delegates to `json.match_schema` which expects objects or marshaled arrays. It silently passes for bare scalars like `42`, so it cannot validate a single string value. The manual `is_string` check is correct here.

**Review by robnester-rh — DISMISSED** [review]
LGTM

File: policy/release/base_image_registries/base_image_registries.rego [hunk 3]
```diff
@@ -107,15 +107,17 @@ _image_ref_permitted(image_ref) if {
...
-_release_public_key := key if {
-	key := rule_data.get(_release_public_key_rule_data_key)
-	is_string(key)
-	key != ""
+_signing_identity := identity if {
+	identities := rule_data.get(_signing_identities_key)
+	is_object(identities)
+	identity := object.get(identities, _signing_identity_name, {})
+	is_object(identity)
+	count(identity) > 0
```
Comment (simonbaird) [inline policy/release/base_image_registries/base_image_registries.rego]: Since we're already doing checks on the rule data format, this could be simpler I think:

```
_signing_identity := rule_data.get(_signing_identities_key)["release"]
```

Comment (simonbaird) [reply policy/release/base_image_registries/base_image_registries.rego]: But we could also consider adding a rule_data rule, e.g.:

```
_signing_identity := rule_data.get_signing_identity("release")
```

File: policy/release/base_image_registries/base_image_registries.rego [hunk 4]
```diff
@@ -105,8 +106,14 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
+	opts := object.union(sigstore.opts, _signing_identity)
```
Comment (simonbaird) [inline policy/release/base_image_registries/base_image_registries.rego]: There was some discussion in slack about whether this object.union makes sense now.

Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Removed. Not sure it provides any value.

**Comment by joejstuart** [comment]
## Response to bot review findings

### Fixed in 786edf41:

- **[edge-case] missing rh-release key** (fullsend) — Added a validation warning when `signing_identities` is a non-empty object but does not contain the expected `rh-release` key.
- **[validation-bypass] prefix validation skipped** (fullsend + Qodo) — `allowed_registry_prefixes` format (type, uniqueness) is now always validated. Only the `minItems: 1` requirement is relaxed when a valid signing identity object is present.
- **[test-inadequate] test_no_release_key_is_noop** (fullsend) — Test now overrides `data.rule_data` explicitly to exercise the prefix-only code path.
- **[scope-drift-naming-mismatch] PR description** (fullsend) — Updated PR description to match the implementation (`signing_identities` with nested `rh-release`).
- **[docs warning code mismatch]** (Qodo, already resolved) — Confirmed the docs correctly use `allowed_registries_provided`.

### By design / intentional:

- **[asymmetric-validation] keyless rekor** — Intentional. Keyless verification fundamentally requires an online Rekor service (the signature is ephemeral and the certificate chain relies on the transparency log for trust). Key-based verification can use an offline `rekor_public_key` or `ignore_rekor` because the key itself provides a durable trust anchor.
- **[authorization-bypass] / [precedence-ordering] prefix matching first** — Intentional migration design. Prefix matching runs first for backward compatibility. The deprecation warning guides users toward signature-based verification. "Preferred" in METADATA refers to the recommended configuration approach, not evaluation order.

### Minor / acknowledged:

- **[naming-convention] removed comment** — The `_rule_data_errors` block now has multiple rules with distinct validation concerns; a single comment wouldn't accurately describe all of them.
- **[error-handling-idiom] error format** — The `signing_identities` validation uses manual type checks because the structure is a nested object, which doesn't fit the `j.validate_schema` pattern used for flat array/string validation. Sigstore opts validation is delegated to `sigstore.validate()`.

**Review by joejstuart — COMMENTED** [review]
Submitting pending review to unblock bot comment replies.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 5]
```diff
@@ -14,17 +14,17 @@ import data.lib.json as j
...
-#   will also validate those images individually.
+#   Verify that the base images used when building a container image are permitted.
+#   Images can be permitted in three ways: by having a valid signature
+#   verified against the `release` entry in `signing_identities` rule data
+#   (preferred), by matching a component digest in the snapshot, or by matching
+#   a registry prefix from `allowed_registry_prefixes` rule data (deprecated).
+#   Registry prefix matching is deprecated and will be removed in a future release.
 # custom:
 #   short_name: base_image_permitted
 #   failure_msg: Base image %q is from a disallowed registry
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Addressed — `failure_msg` updated to `Base image %q is not permitted` (mechanism-agnostic, covers prefix, snapshot, and sigstore paths).

File: policy/release/base_image_registries/base_image_registries.rego [hunk 6]
```diff
@@ -105,8 +105,13 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
+	info := ec.sigstore.verify_image(image_ref, _signing_identity)
+	object.get(info, "success", false) == true
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:116]: Addressed — added a new `base_image_sig_check` deny rule that surfaces individual `info.errors` from `ec.sigstore.verify_image`, following the `source_image.rego` pattern. Operators now see the specific reason verification failed as a separate deny result.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 7]
```diff
@@ -16,15 +16,14 @@ import data.lib.rule_data
...
-#   Base images that are found in the snapshot being validated are also allowed since EC
-#   will also validate those images individually.
+#   Verify that the base images used when building a container image are permitted.
+#   Images can be permitted in three ways: by having a valid signature
+#   verified against the `rh-release` entry in `signing_identities` rule data
+#   (preferred), by matching a component digest in the snapshot, or by matching
+#   a registry prefix from `allowed_registry_prefixes` rule data (deprecated).
+#   Registry prefix matching is deprecated and will be removed in a future release.
 # custom:
 #   short_name: base_image_permitted
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:29]: Addressed — `failure_msg` updated to `Base image %q is not permitted`.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 8]
```diff
@@ -188,4 +199,28 @@ _rule_data_errors contains error if {
...
+	msg := sprintf(
+		"Rule data %s has unexpected format: expected a string, got %s",
+		[_release_key_rule_data_key, type_name(val)],
+	)
+	error := {"message": msg, "severity": "failure"}
+}
+
 _rule_data_key := "allowed_registry_prefixes"
+
+_release_key_rule_data_key := "release_public_key"
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Not applicable — `_release_key_rule_data_key` no longer exists. The constant was replaced by `_signing_identities_key` and `_signing_identity_name` in commit 67309b75.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 9]
```diff
@@ -105,6 +133,22 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
+	key := rule_data.get(_release_key_rule_data_key)
+	is_string(key)
+	key != ""
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Not applicable — in the current code, `ignore_rekor` is not hardcoded. Operators set it directly in the `signing_identities.rh-release` object, which is passed as-is to `ec.sigstore.verify_image`. The old `object.union(sigstore.opts, ...)` approach was removed in commit 67309b75.

File: policy/release/base_image_registries/base_image_registries.rego (hunk 2 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:213]: Not applicable — references `release_public_key` validation which was replaced by `signing_identities` validation using `sigstore.validate()` in commit 0191d638.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 10]
```diff
@@ -105,8 +106,14 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:113]: Not applicable — there is no `object.union` in the current code. `_signing_identity` is passed directly to `ec.sigstore.verify_image` from rule data. The `opts_schema` in `sigstore.rego` has `additionalProperties: false`, so unrecognized keys are caught by `sigstore.validate()`.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 11]
```diff
@@ -170,8 +214,8 @@ _cyclonedx_image_ref(component) := image_ref if {
 	image_ref := sbom.image_ref_from_purl(purl)
 }
 
-# Verify allowed_registry_prefixes is a non-empty list of strings
 _rule_data_errors contains error if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:184]: Not applicable — references `release_public_key` which was refactored away in commit 67309b75 (replaced by `signing_identities`).

File: policy/release/base_image_registries/base_image_registries.rego (hunk 1 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:185]: Not applicable — references `release_public_key` / `not _release_public_key` guard which no longer exists. The current code uses `_prefixes_min_items` to conditionally relax `minItems` while keeping type/uniqueness validation always on.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 12]
```diff
@@ -170,8 +175,8 @@ _cyclonedx_image_ref(component) := image_ref if {
...
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:184]: Not applicable — references a `not _signing_identity` guard on schema validation that does not exist in the current code. See `test_prefix_validation_with_signing_identity` which verifies malformed prefixes are caught even with signing identity configured.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 13]
```diff
@@ -170,8 +177,8 @@ _cyclonedx_image_ref(component) := image_ref if {
 	image_ref := sbom.image_ref_from_purl(purl)
 }
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:182]: Not applicable — same as above. The current code uses `_prefixes_min_items` to conditionally relax `minItems`, not to skip validation entirely. Type/uniqueness checks always run.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 14]
```diff
@@ -105,8 +106,14 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
+	opts := object.union(sigstore.opts, _signing_identity)
+	info := ec.sigstore.verify_image(image_ref, opts)
+	object.get(info, "success", false) == true
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:116]: Not applicable — `sigstore.validate({})` catches the empty object case. `_validate_method` in `sigstore.rego` returns an error because no `public_key` or certificate identity is specified. See `test_signing_identity_validation_integration` for similar validation coverage.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 15]
```diff
@@ -105,8 +105,13 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
+	info := ec.sigstore.verify_image(image_ref, _signing_identity)
+	object.get(info, "success", false) == true
 }
 
+_signing_identity := rule_data.get(_signing_identities_key)[_signing_identity_name]
+
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:120]: Not applicable — same as above. `sigstore.validate({})` produces a failure error via `_validate_method` because no verification method is specified.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 16]
```diff
@@ -170,8 +179,8 @@ _cyclonedx_image_ref(component) := image_ref if {
 	image_ref := sbom.image_ref_from_purl(purl)
 }
 
-# Verify allowed_registry_prefixes is a non-empty list of strings
 _rule_data_errors contains error if {
+	not _release_public_key
 	some e in j.validate_schema(
 		rule_data.get(_rule_data_key),
 		{
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:187]: Not applicable — the current code has no `not _signing_identity` guard on the schema validation rule. Schema validation for `allowed_registry_prefixes` (type, items, uniqueItems) always runs. Only `minItems` is relaxed from 1 to 0 via `_prefixes_min_items` when signing identity is configured. See `test_prefix_validation_with_signing_identity` for coverage.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 17]
```diff
@@ -105,8 +105,13 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:114]: Intentional design — the else-if chain with OR semantics is the migration path from prefix matching to signature-based verification. The rule description (lines 21-27) documents all three methods and the deprecation.

File: antora/docs/modules/ROOT/pages/release_policy.adoc [hunk 18]
```diff
@@ -356,8 +359,8 @@ Rules included:
 * xref:packages/release_git_branch.adoc#git_branch__git_branch[Git branch checks: Builds have a trusted target branch]        
 * xref:packages/release_github_certificate.adoc#github_certificate__gh_workflow_extensions[GitHub Certificate Checks: GitHub Workflow Certificate Extensions]        
 * xref:packages/release_github_certificate.adoc#github_certificate__gh_workflow_name[GitHub Certificate Checks: GitHub Workflow Name]        
-* xref:packages/release_github_certificate.adoc#github_certificate__gh_workflow_ref[GitHub Certificate Checks: GitHub Workflow Repository]        
 * xref:packages/release_github_certificate.adoc#github_certificate__gh_workflow_repository[GitHub Certificate Checks: GitHub Workflow Repository]        
+* xref:packages/release_github_certificate.adoc#github_certificate__gh_workflow_ref[GitHub Certificate Checks: GitHub Workflow Repository]
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply antora/docs/modules/ROOT/pages/release_policy.adoc]: Pre-existing issue on `main`, not introduced by this PR. The `release_policy.adoc` is auto-generated by `make generate-docs`.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 19]
```diff
@@ -105,8 +106,13 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:113]: Intentional design for backward compatibility during migration. The deprecation warning guides operators to adopt `signing_identities`.

File: policy/lib/sigstore/sigstore.rego [hunk 20]
```diff
@@ -11,7 +13,91 @@ default opts := {
...
+	_has_certificate_identity(identity)
+	_has_certificate_oidc_issuer(identity)
+	not _has_value(identity, "rekor_url")
+	error := {"message": "requires rekor_url for keyless verification", "severity": "failure"}
+} else := []
+
+_validate_key_rekor(identity) := [error] if {
+	_has_value(identity, "public_key")
+	object.get(identity, "ignore_rekor", false) != true
+	not _has_value(identity, "rekor_url")
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sigstore/sigstore.rego:85]: Intentional — keyless (Fulcio) verification requires Rekor timestamps for certificate expiry validation. Key-based verification can work without Rekor via `ignore_rekor` or `rekor_public_key`. This is standard sigstore semantics.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 21]
```diff
@@ -188,4 +193,70 @@ _rule_data_errors contains error if {
 	}
 }
 
+_rule_data_errors contains error if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:201]: Intentional — the deprecation warning is the purpose of this change. Warning severity won't block pipelines.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 22]
```diff
@@ -105,8 +106,13 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
+	info := ec.sigstore.verify_image(image_ref, _signing_identity)
+	object.get(info, "success", false) == true
 }
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:118]: Intentional — a warning is emitted when `signing_identities` exists but lacks `rh-release`. Falling back to prefix matching is the correct degraded behavior during migration.

[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:118]: Covered by existing validation — the `not is_object(val)` check catches non-object types for `signing_identities`. If it were an array, `_signing_identity` would be undefined and the policy falls back to prefix matching.

File: antora/docs/modules/ROOT/pages/packages/release_base_image_registries.adoc [hunk 23]
```diff
@@ -43,3 +43,15 @@ Verify the expected information was provided about which base images were used d
...
+=== link:#base_image_registries__registry_prefix_deprecated[Registry prefix matching is deprecated]
+
+Using `allowed_registry_prefixes` to permit base images is deprecated. Configure `release_public_key` to verify base image release signatures instead, which provides stronger cryptographic assurance than registry prefix matching.
+
+*Solution*: Set the `release_public_key` in rule data to enable signature-based base image verification. The key can be an inline PEM-encoded public key or a k8s:// reference to a secret containing the key.
+
+* Rule type: [rule-type-indicator warning]#WARNING#
+* WARNING message: `allowed_registry_prefixes is configured without release_public_key. Migrate to signature-based verification by setting release_public_key in rule data.`
+* Code: `base_image_registries.registry_prefix_deprecated`
+* https://github.com/conforma/policy/blob/{page-origin-refhash}/policy/release/base_image_registries/base_image_registries.rego#L99[Source, window="_blank"]
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply antora/docs/modules/ROOT/pages/packages/release_base_image_registries.adoc]: Not applicable — the current docs have no mention of `registry_prefix_deprecated`. The deprecation warning fires through the existing `allowed_registries_provided` rule, correctly documented. Docs were regenerated with `make generate-docs`.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 24]
```diff
@@ -188,4 +193,70 @@ _rule_data_errors contains error if {
 	}
 }
 
+_rule_data_errors contains error if {
+	prefixes := rule_data.get(_rule_data_key)
+	is_array(prefixes)
+	count(prefixes) > 0
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:204]: Not addressing — minor style preference. The deprecation message intentionally uses different phrasing to distinguish it from schema validation errors.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 25]
```diff
@@ -188,4 +195,43 @@ _rule_data_errors contains error if {
 	}
 }
 
+_rule_data_errors contains error if {
+	prefixes := rule_data.get(_rule_data_key)
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:202]: Not addressing — minor style preference. The deprecation message is clear as-is.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 26]
```diff
@@ -170,7 +176,6 @@ _cyclonedx_image_ref(component) := image_ref if {
 	image_ref := sbom.image_ref_from_purl(purl)
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:181]: Not addressing — the rule body is self-documenting. Comments that just restate what the code does add noise.

File: policy/lib/sigstore/sigstore.rego [hunk 27]
```diff
@@ -11,7 +13,91 @@ default opts := {
...
+
+	# regal ignore:line-length
+	error := {"message": "requires certificate_oidc_issuer or certificate_oidc_issuer_regexp for keyless verification", "severity": "failure"}
+} else := []
+
+_validate_keyless_rekor(identity) := [error] if {
+	not _has_value(identity, "public_key")
+	_has_certificate_identity(identity)
+	_has_certificate_oidc_issuer(identity)
+	not _has_value(identity, "rekor_url")
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sigstore/sigstore.rego:78]: This is not applicable

File: policy/release/base_image_registries/base_image_registries.rego [hunk 28]
```diff
@@ -46,6 +46,28 @@ deny contains result if {
 	result := metadata.result_helper_with_term(rego.metadata.chain(), [image_ref], repo)
 }
 
+# METADATA
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Not applicable — the `base_image_sig_check` rule was removed in 298fdf0b. There is no new deny rule requiring an `effective_on` date.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 29]
```diff
@@ -105,6 +128,20 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:113]: Not applicable — `sigstore.validate(_signing_identity)` is already called in `_rule_data_errors` (line 232). Invalid signing identities produce validation errors before `verify_image` is reached.

File: policy/lib/sigstore/sigstore.rego [hunk 30]
```diff
@@ -11,7 +13,91 @@ default opts := {
...
+	error := {"message": "requires rekor_url for keyless verification", "severity": "failure"}
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sigstore/sigstore.rego:79]: Intentional — keyless verification uses short-lived Fulcio certificates that require Rekor for timestamp proof, so only `rekor_url` is accepted. Key-based verification supports `rekor_url`, `rekor_public_key`, or `ignore_rekor` because the key itself provides the trust anchor.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 31]
```diff
@@ -188,4 +224,70 @@ _rule_data_errors contains error if {
...
+
 _rule_data_key := "allowed_registry_prefixes"
+
+_signing_identities_key := "signing_identities"
+
+_signing_identity_name := "rh-release"
+
+default _prefixes_min_items := 1
+
+_prefixes_min_items := 0 if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:265]: Not applicable — defined as a constant in one place (line 256), which is the correct pattern.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 32]
```diff
@@ -105,6 +128,20 @@ _image_ref_permitted(image_ref) if {
...
+} else if {
+	info := ec.sigstore.verify_image(image_ref, _signing_identity)
+	object.get(info, "success", false) == true
+}
+
+_signing_identity := rule_data.get(_signing_identities_key)[_signing_identity_name]
+
+_sig_verification_errors contains error if {
+	_signing_identity
+	some image_ref in _base_images
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Not applicable — the separate `base_image_sig_check` rule was removed in 298fdf0b. `verify_image` is now only called once (line 110).

File: policy/release/base_image_registries/base_image_registries.rego [hunk 33]
```diff
@@ -188,4 +224,70 @@ _rule_data_errors contains error if {
 	}
 }
 
+_rule_data_errors contains error if {
+	prefixes := rule_data.get(_rule_data_key)
+	is_array(prefixes)
+	count(prefixes) > 0
+	not _signing_identity
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:205]: Not applicable — the deprecation warning has explicit guards: `is_array(prefixes)`, `count(prefixes) > 0`, and `not _signing_identity`. These are distinct error paths from the signing identity validation rules.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 34]
```diff
@@ -170,7 +207,6 @@ _cyclonedx_image_ref(component) := image_ref if {
 	image_ref := sbom.image_ref_from_purl(purl)
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:181]: Not addressing — project convention is no comments unless the why is non-obvious. The validation blocks are self-explanatory.

File: policy/release/base_image_registries/base_image_registries.rego (hunk 33 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:205]: Not addressing — the deprecation message is intentionally a different tone from schema errors. It is a migration nudge, not a validation failure.

File: policy/release/base_image_registries/base_image_registries.rego [hunk 35]
```diff
@@ -105,6 +128,20 @@ _image_ref_permitted(image_ref) if {
 		img := image.parse(component.containerImage)
 	}
 	image.parse(image_ref).digest in allowed_digests
+} else if {
+	info := ec.sigstore.verify_image(image_ref, _signing_identity)
+	object.get(info, "success", false) == true
+}
+
+_signing_identity := rule_data.get(_signing_identities_key)[_signing_identity_name]
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego:119]: Not applicable — `rule_data.get` returns `[]` when absent. Indexing `[]` with `["rh-release"]` is undefined in Rego, so `_signing_identity` is simply undefined, which correctly falls through to prefix matching.

File: policy/release/base_image_registries/base_image_registries.rego (hunk 3 above)
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Already done — line 114 uses `rule_data.get(_signing_identities_key)[_signing_identity_name]` which is the simplified form you suggested.

File: policy/release/base_image_registries/base_image_registries.rego
Comment (st3penta) [inline policy/release/base_image_registries/base_image_registries.rego]: Maybe adapt this to include a mention about the signature checks?

File: policy/release/base_image_registries/base_image_registries.rego [hunk 36]
```diff
@@ -14,20 +14,20 @@ import data.lib.json as j
...
+#   (preferred), by matching a component digest in the snapshot, or by matching
+#   a registry prefix from `allowed_registry_prefixes` rule data (deprecated).
+#   Registry prefix matching is deprecated and will be removed in a future release.
 # custom:
 #   short_name: base_image_permitted
-#   failure_msg: Base image %q is from a disallowed registry
+#   failure_msg: Base image %q is not permitted
 #   solution: >-
 #     Make sure the image used in each task comes from a trusted registry. The list of
 #     trusted registries is a configurable xref:cli:ROOT:configuration.adoc#_data_sources[data source].
```
Comment (st3penta) [inline policy/release/base_image_registries/base_image_registries.rego]: same here, maybe update the solution to match the new logic?

File: policy/lib/sigstore/sigstore.rego [hunk 37]
```diff
@@ -11,7 +13,91 @@ default opts := {
 	"certificate_oidc_issuer_regexp": "",
 	"ignore_rekor": false,
 	"public_key": "",
+	"rekor_public_key": "",
 	"rekor_url": "",
 }
```
Comment (st3penta) [inline policy/lib/sigstore/sigstore.rego:18]: should we document the schema and how it works in some user-facing doc? i'm thinking about surfacing this kind of info: https://redhat-internal.slack.com/archives/C065E0PFN77/p1785444359272129?thread_ts=1785422692.523699&cid=C065E0PFN77

so that we give users clear instructions on how to setup their identities

**Review by st3penta — DISMISSED** [review]
The logic looks good, i'd like the identities schema to be documented, but we can do that later if you want. Feel free to merge

File: policy/release/base_image_registries/base_image_registries.rego (hunk 36 above)
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Updated the solution to describe all three verification methods (signature verification via signing identity, snapshot digest matching, and registry prefix matching) in order of preference.

File: policy/release/base_image_registries/base_image_registries.rego
Comment (joejstuart) [reply policy/release/base_image_registries/base_image_registries.rego]: Updated the solution to mention both signing identity and registry prefix configuration options.

File: policy/lib/sigstore/sigstore.rego (hunk 37 above)
Comment (joejstuart) [reply policy/lib/sigstore/sigstore.rego:18]: Agreed, we'll create a follow-up story to document the signing identities schema and provide clear setup instructions for users.

**Review by st3penta — DISMISSED** [review]

**Review by st3penta — APPROVED** [review]
